contracts/LoanCore.sol

```diff
@@ -224,28 +224,46 @@ contract LoanCore is
      * @param payer                 The party repaying the loan.
      * @param _amountFromPayer      The amount of tokens to be collected from the repayer.
      * @param _amountToLender       The amount of tokens to be distributed to the lender (net after fees).
+     * @param _interestAmount       The interest amount to be paid.
+     * @param _paymentToPrincipal   The portion of the repayment amount that goes to principal.
      */
     function repay(
         uint256 loanId,
         address payer,
         uint256 _amountFromPayer,
-        uint256 _amountToLender
+        uint256 _amountToLender,
+        uint256 _interestAmount,
+        uint256 _paymentToPrincipal
     ) external override onlyRole(REPAYER_ROLE) nonReentrant {
-        LoanLibrary.LoanData memory data = _handleRepay(loanId, _amountFromPayer, _amountToLender);
-
-        // Get promissory notes from two parties involved, then burn
+        (LoanLibrary.LoanData memory data, bool loanRepaid) = _handleRepay(
+            loanId,
+            _amountFromPayer,
+            _amountToLender,
+            _interestAmount,
+            _paymentToPrincipal
+        );
+
+        // get promissory notes from two parties involved
         address lender = lenderNote.ownerOf(loanId);
         address borrower = borrowerNote.ownerOf(loanId);
-        _burnLoanNotes(loanId);
 
-        // Send collected principal + interest, less fees, to lender
+        // if loan is completely repaid burn both notes
+        if (loanRepaid) _burnLoanNotes(loanId);
```

Comment:
Should go down on line 260 next to collateral transfer? We should handle all effects of `loanRepaid == true` in the same place IMO